src/GovUk.Education.ExploreEducationStatistics.Admin/Services/ReleaseDataFileService.cs

```diff
@@ -428,89 +432,144 @@ public async Task<Either<ActionResult, DataFileInfo>> UploadAsZip(Guid releaseVe
                 });
         }
 
-        public async Task<Either<ActionResult, List<DataFileInfo>>> UploadAsBulkZip(
+        public async Task<Either<ActionResult, List<ArchiveDataSetFile>>> ValidateAndUploadBulkZip(
             Guid releaseVersionId,
-            IFormFile bulkZipFormFile)
+            IFormFile zipFile,
+            CancellationToken cancellationToken)
         {
             return await _persistenceHelper
                 .CheckEntityExists<ReleaseVersion>(releaseVersionId)
                 .OnSuccess(_userService.CheckCanUpdateReleaseVersion)
-                .OnSuccess(_ => _dataArchiveValidationService.ValidateBulkDataArchiveFile(
-                        releaseVersionId, bulkZipFormFile)
-                .OnSuccess(async dataArchiveFiles =>
+                .OnSuccessVoid(async _ =>
                 {
-                    var bulkZipFile = new File
+                    var errors = await _dataArchiveValidationService.IsValidZipFile(zipFile);
+                    if (errors.Count > 0)
                     {
-                        CreatedById = _userService.GetUserId(),
-                        RootPath = releaseVersionId,
-                        ContentLength = bulkZipFormFile.Length,
-                        ContentType = bulkZipFormFile.ContentType,
-                        Filename = bulkZipFormFile.FileName.ToLower(),
-                        Type = BulkDataZip,
-                    };
-                    _contentDbContext.Files.Add(bulkZipFile);
-                    await _contentDbContext.SaveChangesAsync();
+                        // TODO: Check this returns
+                        Common.Validators.ValidationUtils.ValidationResult(errors);
+                    }
+                })
+                .OnSuccess(async _ => await _dataArchiveValidationService.ValidateBulkDataArchiveFile(releaseVersionId, zipFile))
+                .OnSuccess(async dataSetFiles =>
+                {
+                    await using var stream = zipFile.OpenReadStream();
+                    var archive = new ZipArchive(stream);
 
-                    // each data/meta file pair are extracted to blob storage by _dataImportService.Import
-                    await UploadFileToStorage(bulkZipFile, bulkZipFormFile);
+                    foreach (var file in dataSetFiles)
```

Comment:
I was expecting us to be uploading just the zip file to temporary storage. I think that probably simplifies things a little as we can rely on `dataImportService#Import` to extract each data set from the zip.

I think the question is how much does this all-files unzipping hurt performance of the `UploadBulkZipDataSetsToTempStorage` endpoint, because I believe right now we're essentially fully decompressing the zip before a response will be returned. If the zip file ends up being in the gigbytes (which I think they will be) then that could be a big performance hit. This could also negatively effect other requests by monopolising the resources of Admin servers.

Previously, the files were being partially decompressed by Admin for validation. I believe that doesn't fully decompress the files - just a sample of each file check the csv files are valid in FileTypeService#IsValidCsvFile.

The actual full data set files decompression was performed by the Data Processor via DataImportService#Import with a `sourceZipFile` specified, which is preferable as it's done by an Azure Function.

I realise all this could represent a fair amount of refactoring, so would be happy to leave this as it is for now if performance isn't a problem - but I suspect it might be with larger zip files.

Reply:
Only saving the archive was the initial plan, but another requirement/preference was to stop referencing the uploaded ZIP files, and only reference the files contained within. The logical step seemed to be to isolate archive file extraction to a single place and only handle the individual files from then onwards.

I'm not sure how to identify or measure if a file is only being partially extracted as you mention, my understanding is that you can't properly read or access a file until it has been fully extracted from an archive, though I may be wrong as it's not something I've looked into before.

The previous extraction behaviour was handled as a separate function, but this still happened as part of the same overall process, i.e. when the user uploads the data file, so they're still sitting around waiting regardless of which backend process is handling the extraction work - I appreciate that the Azure function would have a separate resource allocation which may improve performance though.

Reply:
> Only saving the archive was the initial plan, but another requirement/preference was to stop referencing the uploaded ZIP files, and only reference the files contained within

I think that could still be achieved - i.e. not adding the zip file to the Files table - while still push the decompressing to the Data Processor: we'd need to change `DataImportService#Import` to alternatively read a zip from temporary storage.

> my understanding is that you can't properly read or access a file until it has been fully extracted from an archive

I would hope that when we call `await stream.ReadAsync(sampleBuffer, 0, 1024);` in `IsValidCsvFile` on the archive's stream, it only decompresses those 1024 bytes, not the whole file. But I'm not certain that's correct and could be wrong.

This page seems to suggest that partially extracting a file is possible - although not sure if the ZipArchive implementation will do that: https://superuser.com/questions/1561131/is-it-possible-to-partially-extract-a-zipped-file

I''d be inclined to test your local env with a large bulk zip file - like at least 1 GB, and see how it does.

Reply:
After discussions, the process will remain as-is.